lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -0,0 +1,408 @@
+	// that's backed by the identity private key of the running lnd node.
+	NodeSigner *netann.NodeSigner
+
+	// MaxPaymentMSat is the maximum allowed payment.
+	MaxPaymentMSat lnwire.MilliSatoshi
+
+	// DefaultCLTVExpiry is the default invoice expiry if no values is
+	// specified.
+	DefaultCLTVExpiry uint32
+
+	// ChanDB is a global boltdb instance which is needed to access the
+	// channel graph.
+	ChanDB *channeldb.DB
+}
+
+// AddInvoiceData contains the required data to create a new invoice.
+type AddInvoiceData struct {
+	// An optional memo to attach along with the invoice. Used for record
+	// keeping purposes for the invoice's creator, and will also be set in
+	// the description field of the encoded payment request if the
+	// description_hash field is not being used.
+	Memo string
+
+	// Deprecated. An optional cryptographic receipt of payment which is not
+	// implemented.
+	Receipt []byte
+
+	// The preimage which will allow settling an incoming HTLC payable to
+	// this preimage. If RPreimage is set, RHash should be nil. If both
+	// RPreimage and RHash are nil, a random preimage is generated.
+	Preimage *lntypes.Preimage
+
+	// The hash of the preimage. If RHash is set, RPreimage should be nil.
+	// This condition indicates that we have a 'hold invoice' for which the
+	// htlc will be accepted and held until the preimage becomes known.
+	Hash *lntypes.Hash
+
+	// The value of this invoice in satoshis.
+	Value btcutil.Amount
+
+	// Hash (SHA-256) of a description of the payment. Used if the
+	// description of payment (memo) is too long to naturally fit within the
+	// description field of an encoded payment request.
+	DescriptionHash []byte
+
+	// Payment request expiry time in seconds. Default is 3600 (1 hour).
+	Expiry int64
+
+	// Fallback on-chain address.
+	FallbackAddr string
+
+	// Delta to use for the time-lock of the CLTV extended to the final hop.
+	CltvExpiry uint64
+
+	// Whether this invoice should include routing hints for private
+	// channels.
+	Private bool
+}
+
+// AddInvoice attempts to add a new invoice to the invoice database. Any
+// duplicated invoices are rejected, therefore all invoices *must* have a
+// unique payment preimage.
+func AddInvoice(ctx context.Context, cfg *AddInvoiceConfig,
+	invoice *AddInvoiceData) (*lntypes.Hash, *channeldb.Invoice, error) {
+
+	var paymentPreimage lntypes.Preimage
+	var paymentHash lntypes.Hash
+
+	switch {
+	case invoice.Preimage != nil && invoice.Hash != nil:
+		return nil, nil,
+			errors.New("preimage and hash both set")
+
+	case invoice.Preimage != nil &&
+		bytes.Equal(invoice.Preimage[:], channeldb.UnknownPreimage[:]):
+
+		return nil, nil,
+			fmt.Errorf("cannot use all zeroes as a preimage")
+
+	case invoice.Preimage == nil && invoice.Hash == nil:
+		if _, err := rand.Read(paymentPreimage[:]); err != nil {
+			return nil, nil, err
+		}
+		paymentHash = paymentPreimage.Hash()
+
+	case invoice.Preimage == nil && invoice.Hash != nil:
```

Comment:
for this case, i think we also want to validate that the given hash isn't SHA256(000...), o/w the invoice is technically settled after accept w/o setting the preimage

Reply:
ok, nice case. added